samples/RhythmGame/src/main/cpp/Game.cpp

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2018 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <utils/logging.h>
+#include <thread>
+
+#include "Game.h"
+#include "ui/OpenGLFunctions.h"
+
+Game::Game(AAssetManager *assetManager): mAssetManager(assetManager) {
+}
+
+void Game::start() {
+
+    mClap = SoundRecording::loadFromAssets(mAssetManager, "CLAP.raw");
+    mBackingTrack = SoundRecording::loadFromAssets(mAssetManager, "FUNKY_HOUSE.raw" );
+    mBackingTrack->setPlaying(true);
+    mBackingTrack->setLooping(true);
+
+    mMixer.addTrack(mClap);
+    mMixer.addTrack(mBackingTrack);
+
+    mClapEvents.push(0);
+    mClapEvents.push(24000);
+    mClapEvents.push(48000);
+
+    mClapWindows.push(96000);
+    mClapWindows.push(120000);
+    mClapWindows.push(144000);
+
+    AudioStreamBuilder builder;
+    builder.setFormat(AudioFormat::I16);
```

Comment:
True but chaining the calls makes the code less maintainable because it's slightly more difficult to add and remove stream properties. Is there any advantage, other than not prefixing with `builder.`?

Reply:
Omitting 'builder' removes clutter, so the code is easier to read. If you choose the first parameter to be something that is always needed to be set, and use a line per chained call, then adding / removing setters is no more difficult than in your version. Compare:

builder.setA();
builder.setB();
builder.setC();

builder.setA()
  ->setB()
  ->setC();

[Personally I would prefer setters that return "*this", so it's always "." prefix, but I guess we are stuck with what we have.]

But finally, up to you—there is no style guide for this.

Reply:
When you add a setter you need to remove the trailing `;` from the last line (in this case `->setC();` so there is some marginal extra overhead. Also, I find the difference between the initial `.` and subsequent `->`s to be more visually cluttering (and slightly more confusing: "oh wait, a set method returns a pointer?") than the multiple instances of `builder.`. Not that bothered either way, but will probably leave as is.

Reply:
Sure. I agree that returning a pointer from the setters hurts the looks of client code.